doc/HLD_of_SNS_Client.rst

```diff
@@ -0,0 +1,283 @@
+
+************************
+Logical Specification
+************************
+
+fop builder, NRS and request handler
+========================================
+
+A fop, representing IO operation is created at the VFS or VM entry point1. The fop is then passed to the dummy NRS(23), that immediately passes it down to the request handler. The request handler uses file meta-data to identify the layout and calls layout IO engine to proceed with IO operation.
+
+Layout Schema
+==============
+
+Layout formula generates a parity de-clustered file layout for a particular file, using file id (fid) as an identifier[2]. See Parity De-clustering Algorithm HLD [3] for details. At the moment, m0t1fs supports single file with fid supplied as a mount option.
+
+Layout IO Engine
+==================
+
+Layout IO engine takes as an input layout and a fop (including operation parameters such as file offsets and user data pages). It creates sub-fops3 for individual updates using layout[67] for data[4] , based on pool objects[5]. Sub-fops corresponding to the parity units reference temporarily allocated pages[6], released (under no-cache policy) when processing completes.
+
+ RPC
+=====
+
+Trivial caching policy is used: fops are accumulated[7] in the staging area while IO operation is being processed by the request handler and IO layout engine. Once operation is processed, staged area is un-plugged[8] fops are converted into rpcs[9] and rpcs are transferred to their respective target servers. If IO operation extent is larger than parity group, multiple sibling updates on a given target server are batched together[10].
+
+ Conformance
+=============
+
+- 1[u.FOP] ST
+
+- 2[u.LAYOUT.PARAMETRIZED] ST 
+
+- 3[u.FOP.SNS] ST 
+
+- 4[u.LAYOUT.DATA] ST 
+
+- 5[u.LAYOUT.POOLS] ST 
+
+- 6[u.lib.allocate-page] 
+
+- 7[u.fop.cache.add] 
+
+- 8[u.fop.cache.unplug] 
+
+- 9[u.fop.rpc.to] 
+
+- 10[u.FOP.BATCHING] ST
+
+- [r.sns-client.nocache]: holds per caching policy described in the rpc sub-section.
+
+- [r.sns-client.norecovery]: holds obviously;
+
+- [R.M0.IO.DIRECT]: no-caching and 0-copy for data together implement direct-io.
+
+ Dependencies
+===============
+
+- layout:
+
+  - [u.LAYOUT.SNS] ST: server network striping can be expressed as a layout
+
+  - [u.LAYOUT.DATA] ST: layouts for data are supported
+
+  - [u.LAYOUT.POOLS] ST: layouts use server and device pools for object allocation, location, and identification
+
+  - [u.LAYOUT.PARAMETRIZED] ST: layouts have parameters that are substituted to perform actual mapping
+
+- fop:
+
+  - [u.fop] ST: M0 uses File Operation Packets (FOPs)
```

Comment:
Please use upper case letter or lower case letter in all places. Don't mix them.

Reply:
Incorporated.